// @gen3/guppy/dist/server/middlewares/perIndexTierAccessMiddleware/index.js
"use strict";Object.defineProperty(exports, "__esModule", { value: true });exports.default = void 0;var _config = _interopRequireDefault(require("../../config"));
var _utils = require("../../utils/utils");
var _resolvers = _interopRequireDefault(require("../authMiddleware/resolvers"));
var _resolvers2 = require("../tierAccessMiddleware/resolvers");function _interopRequireDefault(e) {return e && e.__esModule ? e : { default: e };}

const queryTypeMapping = {};
const aggsTypeMapping = {};
const totalCountTypeMapping = {};
let atLeastOneIndexIsRegularAccess = false;

_config.default.esConfig.indices.forEach((item) => {
  if (item.tier_access_level === 'private') {
    queryTypeMapping[item.type] = _resolvers.default;
    aggsTypeMapping[item.type] = _resolvers.default;
  } else if (item.tier_access_level === 'regular') {
    atLeastOneIndexIsRegularAccess = true;
    queryTypeMapping[item.type] = (0, _resolvers2.tierAccessResolver)({
      isRawDataQuery: true,
      esType: item.type,
      esIndex: item.index
    });
    aggsTypeMapping[item.type] = (0, _resolvers2.tierAccessResolver)({ esType: item.type, esIndex: item.index });
    const aggregationName = `${(0, _utils.firstLetterUpperCase)(item.type)}Aggregation`;
    totalCountTypeMapping[aggregationName] = {
      _totalCount: (0, _resolvers2.hideNumberResolver)(true)
    };
  }
  // No additional resolvers necessary for tier_access_level == 'libre'
}, {});

const perIndexTierAccessMiddleware = {
  Query: {
    ...queryTypeMapping
  },
  Aggregation: {
    ...aggsTypeMapping
  },
  ...totalCountTypeMapping
};

if (atLeastOneIndexIsRegularAccess) {
  perIndexTierAccessMiddleware.RegularAccessHistogramForNumber = {
    histogram: (0, _resolvers2.hideNumberResolver)(false),
    asTextHistogram: (0, _resolvers2.hideNumberResolver)(false)
  };

  perIndexTierAccessMiddleware.RegularAccessHistogramForString = {
    histogram: (0, _resolvers2.hideNumberResolver)(false),
    asTextHistogram: (0, _resolvers2.hideNumberResolver)(false)
  };
}var _default = exports.default =

perIndexTierAccessMiddleware;
