// @gen3/guppy/dist/server/middlewares/tierAccessMiddleware/index.js
"use strict";Object.defineProperty(exports, "__esModule", { value: true });exports.default = void 0;var _config = _interopRequireDefault(require("../../config"));
var _utils = require("../../utils/utils");
var _resolvers = require("./resolvers");function _interopRequireDefault(e) {return e && e.__esModule ? e : { default: e };}

// apply this middleware to all es types' data/aggregation resolvers
const queryTypeMapping = {};
const aggsTypeMapping = {};
const totalCountTypeMapping = {};
_config.default.esConfig.indices.forEach((item) => {
  queryTypeMapping[item.type] = (0, _resolvers.tierAccessResolver)({
    isRawDataQuery: true,
    esType: item.type,
    esIndex: item.index
  });
  aggsTypeMapping[item.type] = (0, _resolvers.tierAccessResolver)({ esType: item.type, esIndex: item.index });
  const aggregationName = `${(0, _utils.firstLetterUpperCase)(item.type)}Aggregation`;
  totalCountTypeMapping[aggregationName] = {
    _totalCount: (0, _resolvers.hideNumberResolver)(true)
  };
});
const tierAccessMiddleware = {
  Query: {
    ...queryTypeMapping
  },
  Aggregation: {
    ...aggsTypeMapping
  },
  ...totalCountTypeMapping,
  HistogramForNumber: {
    histogram: (0, _resolvers.hideNumberResolver)(false),
    asTextHistogram: (0, _resolvers.hideNumberResolver)(false)
  },
  HistogramForString: {
    histogram: (0, _resolvers.hideNumberResolver)(false),
    asTextHistogram: (0, _resolvers.hideNumberResolver)(false)
  }
};var _default = exports.default =

tierAccessMiddleware;
