// @gen3/ui-component/dist/components/filters/Chip/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
require("./Chip.css");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
var Chip = function Chip(props) {
  return /*#__PURE__*/_react["default"].createElement("div", {
    className: "g3-chip"
  }, /*#__PURE__*/_react["default"].createElement("div", {
    className: "g3-chip__text"
  }, props.text), /*#__PURE__*/_react["default"].createElement("div", {
    className: "g3-chip__clear-btn",
    role: "button",
    tabIndex: 0,
    onClick: props.onClearButtonClick,
    onKeyPress: props.onClearButtonClick
  }, /*#__PURE__*/_react["default"].createElement("i", {
    className: "g3-icon g3-icon--sm g3-icon-color__lightgray g3-icon--sm g3-icon--cross"
  })));
};
Chip.propTypes = {
  text: _propTypes["default"].node.isRequired,
  onClearButtonClick: _propTypes["default"].func.isRequired
};
var _default = Chip;
exports["default"] = _default;
