// @gen3/guppy/dist/server/logger.js
"use strict";Object.defineProperty(exports, "__esModule", { value: true });exports.default = void 0;var _loglevel = _interopRequireDefault(require("loglevel"));function _interopRequireDefault(e) {return e && e.__esModule ? e : { default: e };}

const originalFactory = _loglevel.default.methodFactory;
_loglevel.default.methodFactory = (methodName, logLevel, loggerName) => {
  const rawMethod = originalFactory(methodName, logLevel, loggerName);

  return (message, ...args) => {
    let timeStr = new Date().toTimeString();
    timeStr = timeStr.substring(0, timeStr.indexOf(' '));
    const combinedMsg = args.reduce((acc, cur) => {
      if (typeof cur === 'string') {
        return `${acc} ${cur}`;
      }
      return `${acc} ${JSON.stringify(cur, null, 4)}`;
    }, message);
    rawMethod(`[${timeStr}] ${methodName.toUpperCase()}: ${combinedMsg}`);
  };
};

const numLevels = {
  0: 'TRACE',
  1: 'DEBUG',
  2: 'INFO',
  3: 'WARN',
  4: 'ERROR',
  5: 'SILENT'
};
_loglevel.default.levelEnums = {
  TRACE: 0,
  DEBUG: 1,
  INFO: 2,
  WARN: 3,
  ERROR: 4,
  SILENT: 5
};

_loglevel.default.setLevel('INFO');
_loglevel.default.setLogLevel = (level) => {
  if (!Object.keys(numLevels).includes(level) && !Object.keys(_loglevel.default.levelEnums).includes(level)) {
    throw new Error(`Invalid log level ${level}`);
  }
  _loglevel.default.setLevel(level);
  _loglevel.default.info('log level set to', numLevels[_loglevel.default.getLevel()]);
};

_loglevel.default.rawOutput = (level, msg) => {
  let parsedLevel = level;
  if (typeof level === 'string') {
    if (!Object.keys(_loglevel.default.levelEnums).includes(level)) {
      throw new Error(`Invalid log level ${level}`);
    }
    parsedLevel = _loglevel.default.levelEnums[level];
  }
  if (parsedLevel >= _loglevel.default.getLevel()) {
    console.log(msg); // eslint-disable-line no-console
  }
};var _default = exports.default =

_loglevel.default;
