// @gen3/guppy/dist/server/es/const.js
"use strict";Object.defineProperty(exports, "__esModule", { value: true });exports.esFieldNumericTextTypeMapping = exports.SCROLL_PAGE_SIZE = exports.NumericTextTypeTypeEnum = exports.AGGS_QUERY_NAME = exports.AGGS_NESTED_QUERY_NAME = exports.AGGS_ITEM_STATS_NAME = exports.AGGS_GLOBAL_STATS_NAME = void 0;const AGGS_QUERY_NAME = exports.AGGS_QUERY_NAME = 'numeric_aggs';
const AGGS_NESTED_QUERY_NAME = exports.AGGS_NESTED_QUERY_NAME = 'numeric_nested_aggs';
const AGGS_GLOBAL_STATS_NAME = exports.AGGS_GLOBAL_STATS_NAME = 'numeric_aggs_stats';
const AGGS_ITEM_STATS_NAME = exports.AGGS_ITEM_STATS_NAME = 'numeric_item_aggs_stats';

const NumericTextTypeTypeEnum = exports.NumericTextTypeTypeEnum = {
  ES_NUMERIC_TYPE: 1,
  ES_TEXT_TYPE: 2
};

const esFieldNumericTextTypeMapping = exports.esFieldNumericTextTypeMapping = {
  keyword: NumericTextTypeTypeEnum.ES_TEXT_TYPE,
  integer: NumericTextTypeTypeEnum.ES_NUMERIC_TYPE,
  long: NumericTextTypeTypeEnum.ES_NUMERIC_TYPE,
  short: NumericTextTypeTypeEnum.ES_NUMERIC_TYPE,
  byte: NumericTextTypeTypeEnum.ES_NUMERIC_TYPE,
  double: NumericTextTypeTypeEnum.ES_NUMERIC_TYPE,
  float: NumericTextTypeTypeEnum.ES_NUMERIC_TYPE,
  half_float: NumericTextTypeTypeEnum.ES_NUMERIC_TYPE,
  scaled_float: NumericTextTypeTypeEnum.ES_NUMERIC_TYPE
};

const SCROLL_PAGE_SIZE = exports.SCROLL_PAGE_SIZE = 10000;
