// @gen3/guppy/dist/components/Utils/filters.js
"use strict";Object.defineProperty(exports, "__esModule", { value: true });exports.updateCountsInInitialTabsOptions = exports.sortTabsOptions = exports.mergeTabOptions = exports.mergeFilters = exports.buildFilterStatusForURLFilter = void 0;var _flat = _interopRequireDefault(require("flat"));
var _lodash = _interopRequireDefault(require("lodash"));function _interopRequireDefault(e) {return e && e.__esModule ? e : { "default": e };}function _typeof(o) {"@babel/helpers - typeof";return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) {return typeof o;} : function (o) {return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;}, _typeof(o);}function ownKeys(e, r) {var t = Object.keys(e);if (Object.getOwnPropertySymbols) {var o = Object.getOwnPropertySymbols(e);r && (o = o.filter(function (r) {return Object.getOwnPropertyDescriptor(e, r).enumerable;})), t.push.apply(t, o);}return t;}function _objectSpread(e) {for (var r = 1; r < arguments.length; r++) {var t = null != arguments[r] ? arguments[r] : {};r % 2 ? ownKeys(Object(t), !0).forEach(function (r) {_defineProperty(e, r, t[r]);}) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) {Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r));});}return e;}function _defineProperty(e, r, t) {return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e;}function _toPropertyKey(t) {var i = _toPrimitive(t, "string");return "symbol" == _typeof(i) ? i : i + "";}function _toPrimitive(t, r) {if ("object" != _typeof(t) || !t) return t;var e = t[Symbol.toPrimitive];if (void 0 !== e) {var i = e.call(t, r || "default");if ("object" != _typeof(i)) return i;throw new TypeError("@@toPrimitive must return a primitive value.");}return ("string" === r ? String : Number)(t);}

/**
   * This function takes two objects containing filters to be applied
   * and combines them into one filter object in the same format.
   * Note: the admin filter takes precedence. Selected values in the user
   * filter will be discarded if the key collides. This is to avoid
   * the user undoing the admin filter. (Multiple user checkboxes increase the
   * amount of data shown when combined, but an admin filter should always decrease
   * or keep constant the amount of data shown when combined with a user filter).
  * */
var mergeFilters = exports.mergeFilters = function mergeFilters(userFilter, adminAppliedPreFilter) {
  var filterAB = _objectSpread({}, userFilter);
  Object.keys(adminAppliedPreFilter).forEach(function (key) {
    if (Object.prototype.hasOwnProperty.call(userFilter, key) &&
    Object.prototype.hasOwnProperty.call(adminAppliedPreFilter, key)) {
      var userFilterSubset = userFilter[key].selectedValues.filter(
        function (x) {return adminAppliedPreFilter[key].selectedValues.includes(x);}
      );
      if (userFilterSubset.length > 0) {
        // The user-applied filter is more exclusive than the admin-applied filter.
        filterAB[key].selectedValues = userFilter[key].selectedValues;
      } else {
        // The admin-applied filter is more exclusive than the user-applied filter.
        filterAB[key].selectedValues = adminAppliedPreFilter[key].selectedValues;
      }
    } else if (Object.prototype.hasOwnProperty.call(adminAppliedPreFilter, key)) {
      filterAB[key] = { selectedValues: adminAppliedPreFilter[key].selectedValues };
    }
  });

  return filterAB;
};

/**
   * This function updates the counts in the initial set of tab options
   * calculated from unfiltered data.
   * It is used to retain field options in the rendering if
   * they are still checked but their counts are zero.
   */
var updateCountsInInitialTabsOptions = exports.updateCountsInInitialTabsOptions = function updateCountsInInitialTabsOptions(
initialTabsOptions,
processedTabsOptions,
filtersApplied,
accessibleFieldCheckList,
allFilterValues)
{
  var updatedTabsOptions = {};
  try {
    // flatten the tab options first
    // {
    //   project_id.histogram: ...
    //   visit.visit_label.histogram: ...
    // }
    var flattenInitialTabsOptions = (0, _flat["default"])(initialTabsOptions, { safe: true });
    var flattenProcessedTabsOptions = (0, _flat["default"])(processedTabsOptions, { safe: true });
    Object.keys(flattenInitialTabsOptions).forEach(function (field) {
      // in flattened tab options, to get actual field name, strip off the last '.histogram' or '.asTextHistogram'
      var actualFieldName = field.replace('.histogram', '').replace('.asTextHistogram', '');

      // check if Filter Value if not skip
      if (!allFilterValues.includes(actualFieldName)) {
        return;
      }

      // possible to have '.' in actualFieldName, so use it as a string
      updatedTabsOptions["".concat(actualFieldName)] = { histogram: [] };
      // if in tiered access mode
      // we need not to process filters for field in accessibleFieldCheckList
      if (accessibleFieldCheckList &&
      accessibleFieldCheckList.includes(actualFieldName) &&
      flattenProcessedTabsOptions["".concat(field)]) {
        updatedTabsOptions["".concat(actualFieldName)].histogram = flattenProcessedTabsOptions["".concat(field)];
        return;
      }
      var histogram = flattenInitialTabsOptions["".concat(field)];
      if (!histogram || _typeof(histogram) !== 'object') {
        console.error("Guppy did not return histogram data for filter field ".concat(actualFieldName)); // eslint-disable-line no-console
        return;
      }
      histogram.forEach(function (opt) {
        var key = opt.key;
        if (typeof key !== 'string') {// key is a range, just copy the histogram
          updatedTabsOptions["".concat(actualFieldName)].histogram = flattenInitialTabsOptions["".concat(field)];
          if (flattenProcessedTabsOptions["".concat(field)] &&
          flattenProcessedTabsOptions["".concat(field)].length > 0 &&
          updatedTabsOptions["".concat(actualFieldName)].histogram) {
            var currentFlattenProcessedTabsOptions = flattenProcessedTabsOptions["".concat(field)][0];

            // if empty count histogram should be removed so filter is not shown
            if (currentFlattenProcessedTabsOptions.count === 0) {
              updatedTabsOptions["".concat(actualFieldName)].histogram = [];
              return;
            }
            updatedTabsOptions["".concat(actualFieldName)].histogram[0].count = currentFlattenProcessedTabsOptions.count;
            var newKey = [0, 0];
            if (currentFlattenProcessedTabsOptions.key[0]) {
              // because of the prefer-destructuring eslint rule
              var newLowerBound = currentFlattenProcessedTabsOptions.key[0];
              newKey[0] = newLowerBound;
            }
            if (currentFlattenProcessedTabsOptions.key[1]) {
              var newHigherBound = currentFlattenProcessedTabsOptions.key[1];
              newKey[1] = newHigherBound;
            }
            updatedTabsOptions["".concat(actualFieldName)].histogram[0].key = newKey;
          }
          return;
        }
        if (flattenProcessedTabsOptions["".concat(field)]) {
          var findOpt = flattenProcessedTabsOptions["".concat(field)].find(function (o) {return o.key === key;});
          if (findOpt) {
            var count = findOpt.count;
            updatedTabsOptions["".concat(actualFieldName)].histogram.push({ key: key, count: count });
          }
        }
      });
      if (filtersApplied["".concat(actualFieldName)]) {
        if (filtersApplied["".concat(actualFieldName)].selectedValues) {
          filtersApplied["".concat(actualFieldName)].selectedValues.forEach(function (optKey) {
            if (!updatedTabsOptions["".concat(actualFieldName)].histogram.find(function (o) {return o.key === optKey;})) {
              updatedTabsOptions["".concat(actualFieldName)].histogram.push({ key: optKey, count: 0 });
            }
          });
        }
      }
    });
  } catch (err) {
    /* eslint-disable no-console */
    // hopefully we won't get here but in case of
    // out-of-index error or obj undefined error
    console.error('error when processing filter data: ', err);
    console.trace();
    /* eslint-enable no-console */
  }
  return updatedTabsOptions;
};

function sortCountThenAlpha(a, b) {
  if (a.count === b.count) {
    return a.key < b.key ? -1 : 1;
  }
  return b.count - a.count;
}

var sortTabsOptions = exports.sortTabsOptions = function sortTabsOptions(tabsOptions) {
  var fields = Object.keys(tabsOptions);
  var sortedTabsOptions = _objectSpread({}, tabsOptions);
  for (var x = 0; x < fields.length; x += 1) {
    var field = fields[x];

    var optionsForThisField = sortedTabsOptions[field].histogram;
    optionsForThisField.sort(sortCountThenAlpha);
    sortedTabsOptions[field].histogram = optionsForThisField;
  }
  return sortedTabsOptions;
};

/**
   * This function takes two TabsOptions object and merge them together
   * The order of merged histogram array is preserved by firstHistogram.concat(secondHistogram)
   */
var mergeTabOptions = exports.mergeTabOptions = function mergeTabOptions(firstTabsOptions, secondTabsOptions) {
  if (!firstTabsOptions || !Object.keys(firstTabsOptions).length) {
    return secondTabsOptions;
  }
  if (!secondTabsOptions || !Object.keys(secondTabsOptions).length) {
    return firstTabsOptions;
  }

  var allOptionKeys = _lodash["default"].union(Object.keys(firstTabsOptions), Object.keys(secondTabsOptions));
  var mergedTabOptions = {};
  allOptionKeys.forEach(function (optKey) {
    if (!mergedTabOptions["".concat(optKey)]) {
      mergedTabOptions["".concat(optKey)] = {};
    }
    if (!mergedTabOptions["".concat(optKey)].histogram) {
      mergedTabOptions["".concat(optKey)].histogram = [];
    }
    var firstHistogram = firstTabsOptions["".concat(optKey)] && firstTabsOptions["".concat(optKey)].histogram ? firstTabsOptions["".concat(optKey)].histogram : [];
    var secondHistogram = secondTabsOptions["".concat(optKey)] && secondTabsOptions["".concat(optKey)].histogram ? secondTabsOptions["".concat(optKey)].histogram : [];
    mergedTabOptions["".concat(optKey)].histogram = firstHistogram.concat(secondHistogram);
  });
  return mergedTabOptions;
};

var buildFilterStatusForURLFilter = exports.buildFilterStatusForURLFilter = function buildFilterStatusForURLFilter(userFilter, tabs) {
  // Converts filter-applied form to filter-displayed form
  // TODO: add support for search filters
  var filteringFields = Object.keys(userFilter);
  var filterStatusArray = tabs.map(function () {return [];});

  for (var tabIndex = 0; tabIndex < tabs.length; tabIndex += 1) {
    var allFieldsForThisTab = _lodash["default"].union(tabs[tabIndex].fields, tabs[tabIndex].asTextAggFields || []);
    filterStatusArray[tabIndex] = allFieldsForThisTab.map(function () {return {};});
    for (var i = 0; i < filteringFields.length; i += 1) {
      var sectionIndex = allFieldsForThisTab.indexOf(filteringFields[i]);
      if (sectionIndex !== -1) {
        var userFilterSmallForm = {};
        var filterVar = userFilter[filteringFields[i]];
        if (_typeof(filterVar) === 'object' && filterVar.selectedValues) {
          // Single select values:
          for (var j = 0; j < filterVar.selectedValues.length; j += 1) {
            userFilterSmallForm[filterVar.selectedValues[j]] = true;
          }
        } else if (_typeof(filterVar) === 'object' && (
        filterVar.lowerBound || filterVar.upperBound)) {
          // Range values:
          userFilterSmallForm = [filterVar.lowerBound, filterVar.upperBound];
        }
        filterStatusArray[tabIndex][sectionIndex] = userFilterSmallForm;
      }
    }
  }

  return filterStatusArray;
};
